File: Dev/C++/Mascaret/src/VEHA/Kernel/Class.cpp
#include "VEHA/Kernel/Class.h"
#include "VEHA/Kernel/Package.h"
#include "VEHA/Kernel/InstanceSpecification.h"
#include "VEHA/Kernel/Property.h"
#include "VEHA/Kernel/Operation.h"
#include "VEHA/Kernel/Property.h"
#include "VEHA/Kernel/InstanceValue.h"
#include "Tools/utils.h"
#include <boost/foreach.hpp>
#include <algorithm>

namespace VEHA
{

Class::Class(const string& name)
:Classifier(name),_package()
{
}

Class::~Class()
{
}

void Class::_addFeaturesToChild(shared_ptr<Classifier> child)
{
	shared_ptr<Class> c(shared_dynamic_cast<Class>(child));
	pair<string,shared_ptr<Property> > p;
	BOOST_FOREACH(p,_attributes)
		c->addAttribute(p.second);
	pair<string,shared_ptr<Operation> > o;
	BOOST_FOREACH(o,_operations)
		c->addOperation(o.second);
}
string Class::_addInstance(InstanceSpecification* value)
{
	return MAP_ADDING(_instances,value);
}
void Class::_removeInstance(InstanceSpecification * value)
{
	_instances.erase(value->getName());
}
void Class::_restockInstance(shared_ptr<InstanceSpecification> value,const string& newName)
{
	_instances.erase(value->getName());
	_instances.insert(make_pair(newName,value.get()));
}

bool Class::addAttribute(shared_ptr<Property> value)
{
	for (size_t i=_children.size();i--;)
		shared_dynamic_cast<Class>(_children[i].lock())->addAttribute(value);
	pair<string,InstanceSpecification* > instance;
	BOOST_FOREACH(instance,_instances)
		instance.second->addSlot(value);
	return _attributes.insert(make_pair(value->getName(),value)).second;
}
bool Class::removeAttribute(const string& name)
{
	map<string,shared_ptr<Property > >::iterator it;
	if ((it=_attributes.find(name))==_attributes.end())
		return false;
	shared_ptr<Property> value =it->second;
	shared_ptr<Property> parentVal;
	for (size_t i=_parent.size();i--;)
	{
		shared_ptr<Class> c(shared_dynamic_cast<Class>(_parent[i].lock()));
		try
		{
			parentVal=MAP_AT(c->getAttributes(),name);
			if (value==parentVal)
				//On ne peut enlever un attribut du parent
				return false;
			else break;
		}
		catch(std::out_of_range&){}
	}
	_attributes.erase(name);
	pair<string,shared_ptr<InstanceSpecification> > instance;
	BOOST_FOREACH(instance,_instances)
	instance.second->removeSlot(name);

	for (size_t i=_children.size();i--;)
	{
		shared_ptr<Class> c(shared_dynamic_cast<Class>(_children[i].lock()));
		shared_ptr<Property> val=MAP_AT(c->getAttributes(),name);
		if (val==value)
			c->removeAttribute(name);
	}
	// parentVal!=null  : set parentVal
	if (parentVal)
		addAttribute(parentVal);
	return true;
}
bool Class::addOperation(shared_ptr<Operation> value)
{
	for (size_t i=_children.size();i--;)
		shared_dynamic_cast<Class>(_children[i].lock())->addOperation(value);
	return _operations.insert(make_pair(value->getName(),value)).second;
}
bool Class::removeOperation(const string &name)
{
	map<string,shared_ptr<Operation > >::iterator it;
	if ((it=_operations.find(name))==_operations.end())
		return false;
	shared_ptr<Operation> value =it->second;
	shared_ptr<Operation> parentVal;
	for (size_t i=_parent.size();i--;)
	{
		shared_ptr<Class> c(shared_dynamic_cast<Class>(_parent[i].lock()));
		try
		{
			parentVal=MAP_AT(c->getOperations(),name);
			if (value==parentVal)
				//On ne peut enlever un attribut du parent
				return false;
			else break;
		}
		catch(std::out_of_range&){}
	}
	_operations.erase(name);

	for (size_t i=_children.size();i--;)
	{
		shared_ptr<Class> c(shared_dynamic_cast<Class>(_children[i].lock()));
		shared_ptr<Operation> val=MAP_AT(c->getOperations(),name);
		if (val==value)
			c->removeOperation(name);
	}
	// parentVal!=null  : set parentVal
	if (parentVal)
		addOperation(parentVal);
	return true;
}

shared_ptr<ValueSpecification> Class::createValueFromString(const string& str)
{
	/*
            map<string, InstanceSpecification* >::iterator it = _instances.find(str);
            if (it != _instances.end())
               return shared_ptr<ValueSpecification>(new InstanceValue(shared_dynamic_cast<InstanceSpecification>(MAP_AT(_instances,str)->shared_from_this())));
            shared_ptr<ValueSpecification
	 */
	map<string,shared_ptr<InstanceSpecification> > instances = getInstances(true);
	return shared_ptr<ValueSpecification>(new InstanceValue(shared_dynamic_cast<InstanceSpecification>(MAP_AT(instances,str)->shared_from_this())));

}

map<string,shared_ptr<InstanceSpecification> > Class::getInstances( bool searchInSubclass) const
{
	map<string,shared_ptr<InstanceSpecification> > result;
	pair<string,InstanceSpecification*> p;
	BOOST_FOREACH(p, _instances)
	{
		result.insert(make_pair<string,shared_ptr<InstanceSpecification> >(p.first,shared_dynamic_cast<InstanceSpecification>(p.second->shared_from_this())));
	}
	if (searchInSubclass)
	{
		for (size_t i=_children.size();i-- ;)
		{
			map<string,shared_ptr<InstanceSpecification> > r=shared_dynamic_cast<Class>(_children[i].lock())->getInstances(true);
			result.insert(r.begin(),r.end());
		}
	}
	return result;
}
shared_ptr<InstanceSpecification> Class::getInstance(const string& name, bool searchInSubclass) const
{
	shared_ptr<InstanceSpecification> result;
	try { result=shared_dynamic_cast<InstanceSpecification>(MAP_AT(_instances,name)->shared_from_this()); }
	catch (std::out_of_range&)
	{
		if (!searchInSubclass) throw;
		for (size_t i=_children.size();i-- && !result;)
		{
			try
			{
				// TODO : dynamic cast pas sur
				result=shared_dynamic_cast<Class>(_children[i].lock())->getInstance(name,searchInSubclass);
			}
			catch (std::out_of_range&)
			{
			}

		}
	}
	if (!result)  throw std::out_of_range("Instance not found");
	else return result;
}
typedef boost::function1<void,InstanceSpecification* > cbInstanceCreate;
void Class::addCallbackOnCreateInstance(const cbInstanceCreate& f)
{
	_cbOnCreate.connect(f);
}

// callback for instances:
void Class::addCallbackOnRemove(const cbSlotValue& f)
{
	_cbOnSlotRemove.connect(f);
}
//		void addCallbackOnRemove(const string& valueName, const cbSlotValue& f);
void Class::addCallbackOnAdd(const cbSlotValue& f)
{
	_cbOnSlotAdd.connect(f);
}
//		void addCallbackOnAdd(const string& valueName, const cbSlotValue& f);
//
//

void Class::fireCBOnCreate(InstanceSpecification* is)
{
	_cbOnCreate(is);
	for (size_t i=_parent.size();i--;)
		shared_dynamic_cast<Class>(_parent[i].lock())->fireCBOnCreate(is);
}
void Class::fireCBOnSlotAdd(const string& slotName,shared_ptr<ValueSpecification> vs)
{
	_cbOnSlotAdd(slotName,vs);
	for (size_t i=_parent.size();i--;)
		shared_dynamic_cast<Class>(_parent[i].lock())->fireCBOnSlotAdd(slotName,vs);
}
void Class::fireCBOnSlotRemove(const string& slotName,shared_ptr<ValueSpecification> vs)
{
	_cbOnSlotRemove(slotName,vs);
	for (size_t i=_parent.size();i--;)
		shared_dynamic_cast<Class>(_parent[i].lock())->fireCBOnSlotRemove(slotName,vs);
}

}

File: Dev/CS/Mascaret/Mascaret/VEHA/Kernel/Class.cs
using System;
using System.Collections.Generic;

//gestion des callBacks (fireCBOnCreate, addCallbackOnRemove ... non implémentées pour le moment)

namespace Mascaret
{
    public class Class : Classifier
    {


        private Package package = null;
        public Package Package
        {
            get { return package; }
            set { package = value; }
        }

        protected List<Connector> connectors = new List<Connector>();
        public List<Connector> getConnectors()
        {
            return connectors;
        }
        public void addConnector(Connector connector)
        {
            connectors.Add(connector);
        }

        private Dictionary<String, Property> attributes;
        public Dictionary<String, Property> Attributes
        {
            get { return attributes; }
            set { attributes = value; }
        }

        public bool addAttribute(Property element)
        {
            if (attributes.ContainsKey(element.name)) return false;
            foreach (Classifier child in Children)
            {
                ((Class)child).addAttribute(element);
            }
            foreach (InstanceSpecification instance in instances.Values)
            {
                instance.addSlot(element);
            }
            attributes.Add(element.name, element);
            return true;
        }

        public bool removeAttribute(string name)
        {
            if (attributes.ContainsKey(name))
            {
                Property val = attributes[name];
                Property parentVal = null;
                foreach (Classifier parent in Parents)
                {
                    Class c = (Class)parent;
                    if (c.Attributes.ContainsKey(name))
                    {
                        parentVal = c.Attributes[name];
                        if (val == parentVal)
                            return false;
                        else break;
                    }
                }

                Attributes.Remove(name);

                foreach (InstanceSpecification instance in instances.Values)
                {
                    instance.removeSlot(name);
                }

                foreach (Classifier child in Children)
                {
                    Class c = (Class)child;
                    Property otherVal;
                    if (c.Attributes.ContainsKey(name))
                    {
                        otherVal = c.Attributes[name];
                        if (otherVal == val)
                            c.removeAttribute(name);
                    }
                }

                if (parentVal != null)
                    addAttribute(parentVal);
                return true;

            }
            else
                return false;
        }

        public virtual bool hasAttribute(string name)
        {
            return attributes.ContainsKey(name);
        }


        private Dictionary<String, Operation> operations;
        public Dictionary<String, Operation> Operations
        {
            get { return operations; }
            set { operations = value; }
        }

        public bool addOperation(Operation element)
        {
            if (operations.ContainsKey(element.name)) return false;
            foreach (Classifier child in Children)
            {
                ((Class)child).addOperation(element);
            }
            
            operations.Add(element.name, element);
            return true;
        }

        public bool removeOperation(string name)
        {
            if (operations.ContainsKey(name))
            {
                Operation val = operations[name];
                Operation parentVal = null;
                foreach (Classifier parent in Parents)
                {
                    Class c = (Class)parent;
                    if (c.Operations.ContainsKey(name))
                    {
                        parentVal = c.Operations[name];
                        if (val == parentVal)
                            return false;
                        else break;
                    }
                }

                Operations.Remove(name);

                foreach (Classifier child in Children)
                {
                    Class c = (Class)child;
                    Operation otherVal;
                    if (c.Operations.ContainsKey(name))
                    {
                        otherVal = c.Operations[name];
                        if (otherVal == val)
                            c.removeOperation(name);
                    }
                }

                if (parentVal != null)
                    addOperation(parentVal);
                return true;

            }
            else
                return false;
        }

        public virtual bool hasOperation(string name)
        {
            return operations.ContainsKey(name);
        }


        private Dictionary<String, InstanceSpecification> instances;
        public Dictionary<String, InstanceSpecification> Instances
        {
            get { return instances; }
            set { instances = value; }
        }

        public Class(string name)
            : base(name)
        {
            attributes = new Dictionary<string, Property>();
            instances = new Dictionary<string, InstanceSpecification>();
            operations = new Dictionary<string, Operation>();

        }

        public override string getFullName()
        {
            if (package != null)
                return package.getFullName() + "::" + name;
            else
                return name;
        }

        public override ValueSpecification createValueFromString(string str)
        {
            Dictionary<string, InstanceSpecification> instances = getInstances(true);
            //Debug.Log (" Class :: CreateValueFromString : " + getFullName() + " / " + str +" : " + instances.Count);
            InstanceValue curInstanceValue = new InstanceValue((InstanceSpecification)(instances[str]));

            return (ValueSpecification)curInstanceValue;
        }

        public void restockInstance(InstanceSpecification specValue, string newName)
        {
            instances.Remove(specValue.name);
            instances.Add(newName, specValue);
        }

        public Dictionary<string, InstanceSpecification> getInstances(bool searchInSubclass)
        {
            Dictionary<string, InstanceSpecification> result = new Dictionary<string, InstanceSpecification>();
            foreach (KeyValuePair<string, InstanceSpecification> p in instances)
            {
                result.Add(p.Key, p.Value);
            }
            if (searchInSubclass)
            {
                for (int i = 0; i < Children.Count; i++)
                {
                    foreach (KeyValuePair<string, InstanceSpecification> r in ((Class)Children[i]).getInstances(true))
                    {
                        result.Add(r.Key, r.Value);
                    }
                }
            }
            return result;
        }

        protected override void _addFeaturesToChild(Classifier child)
        {
            Class c = (Class)child;
            foreach (Property p in attributes.Values)
            {
                c.addAttribute(p);
            }
            /*
            foreach(Operation op in operations.ValuesList)
            {
                c.addOperation(op);
            }*/
        }

        public InstanceSpecification getInstance(string name, bool searchinSubclass)
        {
            InstanceSpecification result;
            if (instances.ContainsKey(name))
                return instances[name];
            else
            {
                if (searchinSubclass)
                {
                    foreach (Classifier child in Children)
                    {
                        result = ((Class)child).getInstance(name, searchinSubclass);
                        if (result != null)
                            return result;
                    }
                }
            }
            return null;
        }

        public void print()
        {
            System.Console.WriteLine(getFullName());

            if (OwnedBehavior != null) System.Console.WriteLine("Behaviors : " + OwnedBehavior.Keys.Count);
            /*
            for (int i = 0; i < OwnedBehavior.KeysList.Count; i++)
                Debug.Log(OwnedBehavior.KeysList[i]);
                */
        }


    }
}
